### PR TITLE
add delete sealer sub command

### PR DESCRIPTION
**Pull request by x-shadow-man**
### Describe what this PR does / why we need it
add clean sub command for sealer. It can delete clustfile ip list from command. for example: 
sealer clean --masters x.x.x.x --nodes x.x.x.x

### CODE REVIEW

**Comment by allencloud** [comment]
I am wondering if `clean` is the explicit and exactly matching word to represent the command. 

Word clean is usually used to erase things, especially static things. While here, I wish to make sure if we need the function of disabling some kubernetes nodes and picking them from the original cluster. If so, I am afraid that word delete would be better. @x-shadow-man @fanux

File: apply/clean.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+	applier, err := NewApplier(cluster)
+	if err != nil {
+		logger.Error("failed to init applier, err: %s", err)
+		return nil
+	}
+	return applier
+}
+
+func returnFilteredIPList(clusterIPList []string, toBeDeletedIPList []string) []string {
```
Comment (allencloud) [inline apply/clean.go:82]: I wish that you should not modify the input slice. I do not think it is a good way to define and implement your function.

**Comment by fanux** [comment]
@allencloud Yes, maybe `delete` is better? Sealer already has `sealer delete -f Clusterfile` command, I think we can distinguish `-f -m -n` flag.

like:
```
sealer delete -f Clusterfile # delete all nodes and clean the cluster.
sealer delete -m x.x.x.x # delete master
sealer delete -n x.x.x.x # delete node
```

@x-shadow-man

**Comment by allencloud** [comment]
> @allencloud Yes, maybe `delete` is better? Sealer already has `sealer delete -f Clusterfile` command, I think we can distinguish `-f -m -n` flag.
> 
> like:
> 
> ```
> sealer delete -f Clusterfile # delete all nodes and clean the cluster.
> sealer delete -m x.x.x.x # delete master
> sealer delete -n x.x.x.x # delete node
> ```
> 
> @x-shadow-man

I totally on the delete command. Actually delete all the nodes and delete partial nodes could be regarded as one kind of same action.

**Comment by fanux** [comment]
@allencloud this command will uninstall kubernetes on a node or master, and clean all rootfs files, so I think clean is fine.  But we need merge clean and delete command.

File: apply/clean.go (hunk 1 above)
Comment (x-shadow-man) [reply apply/clean.go]: Yeah, thx, u'r right. i'll fix it.

Comment (x-shadow-man) [reply apply/clean.go]: > I wish that you should not modify the input slice. I do not think it is a good way to define and implement your function.
Yeah, thx, u'r right. i'll fix it.

File: utils/clusterfile_util.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	"strings"
+
+	"github.com/alibaba/sealer/cert"
+	"github.com/alibaba/sealer/logger"
+
+	"io/ioutil"
+	"os"
+)
+
+func GetDefaultClusterFilePath(clusterFile string) (string, error) {
```
Comment (fanux) [inline utils/clusterfile_util.go:14]: https://github.com/alibaba/sealer/blob/main/common/common.go#L136  The common module already has those functions

Comment (x-shadow-man) [reply utils/clusterfile_util.go:14]: ok，i'll read first. thx

File: apply/scale.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	cluster := &v1.Cluster{}
+	if err := utils.UnmarshalYamlFile(clusterfile, cluster); err != nil {
+		logger.Error("clusterfile parsing failed, please check:", err)
+		return nil
+	}
+	if scalingArgs.Nodes == "" && scalingArgs.Masters == "" {
+		logger.Error("The node or master parameter was not committed")
+		return nil
+	}
+	if cluster.Spec.Provider == "BAREMETAL" {
```
Comment (fanux) [inline apply/scale.go:38]: I think that logic is a litter complex，may be using a function is better.

```
func scaleBareMetal(...)...{
    if cluster.Spec.Provider != common. BAREMETAL {
        return ...
   }
}
```

File: sealer/cmd/delete.go [hunk 4]
```diff
@@ -52,20 +81,35 @@ var deleteCmd = &cobra.Command{
...
-			os.Exit(1)
+		if deleteArgs.Nodes != "" || deleteArgs.Masters != "" {
+			applier := apply.NewScalingApplierFromArgs(deleteClusterFile, deleteArgs, false)
+			if applier == nil {
+				os.Exit(1)
+			}
+			if err := applier.Apply(); err != nil {
+				logger.Error(err)
+				os.Exit(1)
+			}
```
Comment (fanux) [inline sealer/cmd/delete.go:92]: return fast.

```
if xxx {
    xxx
    return
}
xxxx
```

when the `else` is end of the function, we can return in `if`

File: sealer/cmd/delete.go [hunk 5]
```diff
@@ -52,20 +81,35 @@ var deleteCmd = &cobra.Command{
...
-		if err != nil {
-			logger.Error(err)
-			os.Exit(1)
-		}
-		if err = applier.Delete(); err != nil {
-			logger.Error(err)
-			os.Exit(1)
+		if deleteArgs.Nodes != "" || deleteArgs.Masters != "" {
+			applier := apply.NewScalingApplierFromArgs(deleteClusterFile, deleteArgs, false)
+			if applier == nil {
```
Comment (fanux) [inline sealer/cmd/delete.go:86]: needs err log here

File: sealer/cmd/delete.go [hunk 6]
```diff
@@ -20,23 +20,52 @@ import (
...
+	sealer delete --masters 2 --nodes 3 -f /root/.sealer/specify-cluster/Clusterfile
+delete all:
+	sealer delete --all [--force]
+	sealer delete -f /root/.sealer/mycluster/Clusterfile [--force]
+`,
 	Run: func(cmd *cobra.Command, args []string) {
 		force, err := cmd.Flags().GetBool("force")
 		if err != nil {
 			logger.Error(err)
 			os.Exit(1)
```
Comment (fanux) [inline sealer/cmd/delete.go:53]: too many os.Exit in this function, I think you can put all the logic into one function and return error. So Run will much simpler, just needs one os.Exit(1).

File: apply/shrink.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+			}
+			return nil
+		}
+		return fmt.Errorf(" Parameter error: The number of join masters or nodes that must be submitted to use cloud service！")
+	default:
+		return fmt.Errorf(" clusterfile provider type is not found ！")
+	}
+}
+
+func returnFilteredIPList(clusterIPList []string, toBeDeletedIPList []string) []string {
```
Comment (fanux) [inline apply/shrink.go:88]:
```suggestion
func returnFilteredIPList(clusterIPList []string, toBeDeletedIPList []string) (res []string) {
    for _,ip := range clusterIPList {
        if utils.NotIn(ip,toBeDeletedIPList) {
            res = append(res,ip)
        }
    }
    return 
}
```

File: apply/expand.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		return fmt.Errorf(" Parameter error: The number of join masters or nodes that must be submitted to use cloud service！")
+	default:
+		return fmt.Errorf(" clusterfile provider type is not found ！")
+	}
+}
+
+func StrToInt(str string) int {
+	num, err := strconv.Atoi(str)
+	if err != nil {
+		logger.Error("String to digit conversion failed:", err)
```
Comment (fanux) [inline apply/expand.go:83]: errors occurred but return 0, the caller will set master and nodes count as 0, this is horrible.

File: apply/shrink.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/alibaba/sealer/common"
+	v1 "github.com/alibaba/sealer/types/api/v1"
+)
+
+type Shrink struct {
```
Comment (fanux) [inline apply/shrink.go:26]: Shrink and Expand is a verb, it should be a function, not struct I think.

File: apply/shrink.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+func (p *Shrink) Scale(cluster *v1.Cluster, scalingArgs *common.RunArgs) error {
+	switch cluster.Spec.Provider {
+	case common.BAREMETAL:
+		if err := PreProcessIPList(scalingArgs); err != nil {
+			return err
+		}
+		if (!IsIPList(scalingArgs.Nodes) && scalingArgs.Nodes != "") || (!IsIPList(scalingArgs.Masters) && scalingArgs.Masters != "") {
```
Comment (fanux) [inline apply/shrink.go:35]: The logic can be simpler, like this:
```suggestion
if IsIPList(nodes) && nodes != "" {
...
   cluster.spec.nodes = removeIPListDuplicatesAndEmpty(nodes)
}
if IsIPList(masters) && masters != "" {
...
    cluster.spec.master = removeIPListDuplicatesAndEmpty(nodes)
}
```

Comment (x-shadow-man) [reply apply/shrink.go:35]: adopt！

**Review by fanux — APPROVED** [review]

**Pull request merged**